dedalus/core/timesteppers.py

```diff
@@ -727,3 +731,282 @@ class RKGFY(RungeKuttaIMEX):
+            spX = sp.gather_inputs(state_fields)
+            apply_sparse(sp.M_min, spX, axis=0, out=MX0.get_subdata(sp))
+            apply_sparse(sp.L_min, spX, axis=0, out=LX0.get_subdata(sp))
+
+        # Compute stages
+        # (M + k Hii L).X(n,i) = M.X(n,0) + k Aij F(n,j) - k Hij L.X(n,j)
+        for i in range(1, self.stages + 1):
+            # Compute L.X(n,i-1), already done for i=1
+            if i > 1:
+                LXi = LX[i-1]
+                # Ensure coeff space before subsystem gathers
+                evaluator.require_coeff_space(state_fields)
+                for sp in subproblems:
+                    spX = sp.gather_inputs(state_fields)
+                    apply_sparse(sp.L_min, spX, axis=0, out=LXi.get_subdata(sp))
+
+            # Compute F(n,i-1), only doing output on first evaluation
+            if i == 1:
+                evaluator.evaluate_scheduled(iteration=iteration, wall_time=wall_time, sim_time=solver.sim_time, timestep=dt)
+            else:
+                evaluator.evaluate_group('F')
+            Fi = F[i-1]
+            for sp in subproblems:
+                # F fields should be in coeff space from evaluator
+                sp.gather_outputs(F_fields, out=Fi.get_subdata(sp))
+
+            # Construct RHS(n,i)
+            if RHS.data.size:
+                np.copyto(RHS.data, MX0.data)
+                for j in range(0, i):
+                    # RHS.data += (k * A[i,j]) * F[j].data
+                    axpy(a=(k * A[i, j]), x=F[j].data, y=RHS.data)
+                    # RHS.data -= (k * H[i,j]) * LX[j].data
+                    axpy(a=-(k * H[i, j]), x=LX[j].data, y=RHS.data)
+
+            # Solve for stage (Compute X)
+            k_Hii = k * H[i, i]
+            # Ensure coeff space before subsystem scatters
+            for field in state_fields:
+                field.preset_layout('c')
+            for sp in subproblems:
+                # Construct LHS(n,i) using old H
+                if update_LHS:
+                    if STORE_EXPANDED_MATRICES:
+                        # sp.LHS.data[:] = sp.M_exp.data + k_Hii*sp.L_exp.data
+                        np.copyto(sp.LHS.data, sp.M_exp.data)
+                        axpy(a=k_Hii, x=sp.L_exp.data, y=sp.LHS.data)
+                    else:
+                        sp.LHS = (sp.M_min + k_Hii * sp.L_min)  # CREATES TEMPORARY
+                    sp.LHS_solvers[i] = solver.matsolver(sp.LHS, solver)
+                # Slice out valid subdata, skipping invalid components
+                spRHS = RHS.get_subdata(sp)
+                spX = sp.LHS_solvers[i].solve(spRHS)  # CREATES TEMPORARY
+                sp.scatter_inputs(spX, state_fields)
+
+            # If this is the last iteration, swap H's last row with b_hat and compute X_hat
+            if i == self.stages-1:
+                # Swap the last row of H with b_hat
+                H[-1, :] = b_hat_IM
+                # A[-1, :] = b_hat_EX <- not sure if this is needed
```

Comment:
This is interesting. I have a few questions:
 - Did you check that you get the desired order of accuracy for both methods? I understand that you cannot just sum because of algebraic constraints. But if you solve, you are doing something different. If you get the correct order, is it by chance or is it expected?
 - Why did you comment out the above line? I would expect that it is needed in general. 
 - How many more factorisations do you need with this scheme? The RK methods so far implemented in Dedalus are all singly diagonally implicit, such that you need only one factorisation per step. But I assume $\hat{b}_s \neq H_{ss}$, such that you need one more factorisation. If the step size is kept constant, do you need to factorize twice again in the next step or is there a cache?